zebra-chain/src/parallel/tree.rs

```diff
@@ -160,36 +191,106 @@ impl NoteCommitmentTrees {
     }
 
     /// Update the sapling note commitment tree.
+    #[allow(clippy::unwrap_in_result)]
     fn update_sapling_note_commitment_tree(
         mut sapling: Arc<sapling::tree::NoteCommitmentTree>,
         sapling_note_commitments: Vec<sapling::tree::NoteCommitmentUpdate>,
-    ) -> Result<Arc<sapling::tree::NoteCommitmentTree>, NoteCommitmentTreeError> {
+    ) -> Result<
+        (
+            Arc<sapling::tree::NoteCommitmentTree>,
+            Option<(sapling::tree::Node, bool, u16)>,
+        ),
+        NoteCommitmentTreeError,
+    > {
         let sapling_nct = Arc::make_mut(&mut sapling);
+        let distance_to_complete_subtree =
+            (1 << TRACKED_SUBTREE_HEIGHT) - sapling_nct.count() % (1 << TRACKED_SUBTREE_HEIGHT);
+        let should_find_subtree_roots =
+            sapling_note_commitments.len() as u64 >= distance_to_complete_subtree;
+
+        // MAX_BLOCK_BYTES is less than 2^21 bytes, notes are 2^5 bytes, and a full subtree has
+        // 2^16 notes. Blocks should not contain multiple level 16 roots.
+        let mut subtree_root = None;
 
-        for sapling_note_commitment in sapling_note_commitments {
-            sapling_nct.append(sapling_note_commitment)?;
+        if should_find_subtree_roots {
+            let starts_new_subtree =
+                sapling_note_commitments.len() as u64 != distance_to_complete_subtree;
+            let complete_subtree_note_in_block_idx = distance_to_complete_subtree as usize - 1;
+            for (note_idx, sapling_note_commitment) in
+                sapling_note_commitments.into_iter().enumerate()
+            {
+                sapling_nct.append(sapling_note_commitment)?;
+                if note_idx == complete_subtree_note_in_block_idx {
+                    subtree_root = Some((
+                        sapling_nct
+                            .subtree_root::<TRACKED_SUBTREE_HEIGHT>()
+                            .expect("should be frontier after appending a note"),
+                        starts_new_subtree,
+                        sapling_nct.subtree_index::<TRACKED_SUBTREE_HEIGHT>() as u16,
```

Comment:
This `as` conversion could be incorrect if the underlying function returns a value greater than `u16::MAX`.

I'd suggest doing this conversion inside `subtree_index()`, and using `try_into().expect("always within u16")` to 
avoid accidental truncation.